File: data-structures/trie/trie.js
class TrieNode {
    constructor() {
        this.child = {};
        this.endOfWord = false;
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode()

    }

    find(word) {
        let root = this.root;
        for (let char of word) {

            root = root.child[char]
        }

        return root ? root.endOfWord : false
    }

    findPrefix(word) {
        let root = this.root;
        let found = false;
        for (let char of word) {
            if (!root.child[char]) {
                found = false;
                break;
            } else {
                found = true;
            }
            root = root.child[char]
        }
        return found
    }

    insert(word) {
        let root = this.root;

        for (let char of word) {
            if (!root.child[char]) {
                root.child[char] = new TrieNode()

            }

            root = root.child[char]

        }
        root.endOfWord = true

    }

    delete(word) {

        this.root.child = this.deleteHelper(word, this.root.child, 0)

    }

    deleteHelper(word, current, idx) {
       
        const newCurrent = current[word[idx]]
        if(!newCurrent){
           return false 
        }
        const next = newCurrent.child

        if (idx === word.length - 1) {
            if (newCurrent.endOfWord) {
                newCurrent.endOfWord = false;
            }
            if (Object.keys(newCurrent.child).length === 0) {
                return {}
            } else {

                return current;
            }

        }
        const returnedResult = this.deleteHelper(word, next, idx + 1)
        current[word[idx]].child = returnedResult;
        if (current[word[idx]].endOfWord === false && Object.keys(current[word[idx]].child).length === 0) {

            return {}
        }

        return current;
    }

}

var trie = new Trie();

trie.insert("hello")
trie.insert("hell")
trie.delete("hell")

File: data-structures/trie/trie.go
package main

import "fmt"

type TrieNode struct {
	child     map[rune]*TrieNode
	endOfWord bool
}

type Trie struct {
	root *TrieNode
}

func (t *Trie) insert(word string) {
	root := t.root

	for _, char := range word {
		if root.child[char] == nil {
			root.child[char] = &TrieNode{make(map[rune]*TrieNode), false}
		}

		root = root.child[char]

	}
	root.endOfWord = true

}

func (t *Trie) find(word string) bool {
	root := t.root
	for _, char := range word {
		value := root.child[char]
		if value == nil {
			return false

		} else {
			root = value
		}

	}

	if root != nil {
		return root.endOfWord
	}
	return false

}

func (t *Trie) findPrefix(word string) bool {
	root := t.root
	for _, char := range word {
		value := root.child[char]
		if value == nil {
			return false

		} else {
			root = value
		}

	}

	if root != nil {
		return true
	}
	return false

}

func (t *Trie) delete(word string) {
	t.root.child = t.deleteHelper(word, t.root.child, 0)
}

func (t *Trie) deleteHelper(word string, current map[rune]*TrieNode, idx int) map[rune]*TrieNode {
	newCurrent := current[rune(word[idx])]
	if newCurrent == nil {
		return nil
	}
	next := newCurrent.child

	if idx == len(word)-1 {
		if newCurrent.endOfWord == true {
			newCurrent.endOfWord = false
		}
		if len(newCurrent.child) == 0 {
			return nil
		} else {

			return current
		}

	}
	returnedResult := t.deleteHelper(word, next, idx+1)
	current[rune(word[idx])].child = returnedResult
	if current[rune(word[idx])].endOfWord == false && len(current[rune(word[idx])].child) == 0 {

		return nil
	}

	return current
}

func (t *Trie) create() {
	t.root = &TrieNode{make(map[rune]*TrieNode), false}
}

func main() {
	trie := Trie{}
	trie.create()
	trie.insert("hello")
	trie.insert("helloworld")

	trie.delete("hello")

	fmt.Println(trie.find("helloworld"))

}
